firmware/controllers/engine_cycle/spark_logic.cpp

```diff
@@ -466,13 +476,29 @@ static void prepareIgnitionSchedule() {
 }
 
 static void scheduleAllSparkEventsUntilNextTriggerTooth(uint32_t trgEventIndex, efitick_t edgeTimestamp) {
-	AngleBasedEvent *current, *tmp;
+	AngleBasedEvent *current, *tmp, *keephead;
+	AngleBasedEvent *keeptail = nullptr;
+
+	{
+		chibios_rt::CriticalSectionLocker csl;
+
+		keephead = engine->angleBasedEventsHead;
+		engine->angleBasedEventsHead = nullptr;
+	}
 
-	LL_FOREACH_SAFE2(engine->angleBasedEventsHead, current, tmp, nextToothEvent)
+	LL_FOREACH_SAFE2(keephead, current, tmp, nextToothEvent)
 	{
 		if (current->position.triggerEventIndex == trgEventIndex) {
 			// time to fire a spark which was scheduled previously
-			LL_DELETE2(engine->angleBasedEventsHead, current, nextToothEvent);
+
+			// Yes this looks like O(n^2), but that's only over the entire engine
+			// cycle.  It's really O(mn + nn) where m = # of teeth and n = # events
+			// fired per cycle.  The number of teeth outweigh the number of events, at
+			// least for 60-2....  So odds are we're only firing an event or two per
+			// tooth, which means the outer loop is really only O(n).  And if we are
+			// firing many events per teeth, then it's likely the events before this
+			// one also fired and thus the call to LL_DELETE2 is closer to O(1).
```

Comment:
in addition, the sum M + N is roughly proportional to the number of cylinders: engines with fewer trigger teeth will have more events per tooth, and engines with many teeth will only have an event or two per tooth. I'm fine that this looks O(N^2), since it's really close to O(1)